File: spsc-lite-haskell/MSG.hs
-- Most Specific Generalization
module MSG where

import qualified Data.Map as Map
import Control.Monad.State

import SParsers
import SLanguage
import ShowUtil
import Algebra

data Gen = Gen Exp Subst Subst
             deriving Eq

msg :: Exp -> Exp -> State Int Gen

msg e1 e2 =
  do k <- freshName
     let u = Gen (Var k) (Map.singleton k e1) (Map.singleton k e2) in
       msgLoop u

msgLoop :: Gen -> State Int Gen

msgLoop u =
  do let u' = mergeSubexp u
     u'' <- commonFunctor u'
     if u'' == u then return u
                 else msgLoop u''

commonFunctor :: Gen -> State Int Gen

commonFunctor u @ (Gen e m1 m2) =
  case [(k, e1) | (k, e1) <- Map.assocs m1, theSameFunctor e1 (m2 Map.! k)] of
    [] -> do { return u}
    (k, e1@(Call ctr name args1)) : _ ->
      let e2@(Call _ _ args2) = m2 Map.! k in
      do ns <- freshNameList (length args1)
         let vs = map Var ns
             e' = applySubst (Map.singleton k (Call ctr name vs)) e
             m1' = (Map.delete k m1) `Map.union` Map.fromList (ns `zip` args1)
             m2' = (Map.delete k m2) `Map.union` Map.fromList (ns `zip` args2) in
           return $ Gen e' m1' m2'


mergeableKeyPairs m1 m2 =
  [ (k1,k2) | (k1,e1) <- Map.assocs m1, (k2,e2) <- Map.assocs m1, k1 < k2,
               e1 == e2, m2 Map.! k1 == m2 Map.! k2]

mergeSubexp u @ (Gen e m1 m2) =
  case mergeableKeyPairs m1 m2 of
    [] -> u
    (k1, k2) : _ ->
      let e' = applySubst (Map.singleton k1 (Var k2)) e
          m1' = Map.delete k1 m1
          m2' = Map.delete k1 m2
      in Gen e' m1' m2'

---- Show

instance Show Gen where
    show (Gen e subst1 subst2) =
      show e ++ " =>> "
        ++ "{" ++ showBindings (Map.toAscList subst1) ++ "}"
        ++ "{" ++ showBindings (Map.toAscList subst2) ++ "}" 
  

File: spsc-lite-julia/src/msg.jl
module MSG

using SPSC.SLanguage
using SPSC.SParsers
using SPSC.ShowUtil
using SPSC.Algebra

struct Gen
    e::Exp
    s1::Subst
    s2::Subst
end

function Base.show(io::IO, gen::Gen)
    print(io, "")
    print(io, gen.e)
    print(io, " =>> {")
    print(io, substToString(gen.s1))
    print(io, "}{")
    print(io, substToString(gen.s2))
    print(io, "}")
end

function commonFunctor!(ng::NameGen, e::Exp, s1::Subst, s2::Subst)::Exp
    for n in keys(s1)
        e1 = s1[n]
        e2 = s2[n]
        (e1 isa CFG && e2 isa CFG && theSameFunctor(e1, e2)) || continue
        c1 = e1::CFG
        c2 = e2::CFG
        l = length(c1.args)
        ns = freshNameList(ng, l)
        delete!(s1, n)
        merge!(s1, Subst(zip(ns, c1.args)))
        delete!(s2, n)
        merge!(s2, Subst(zip(ns, c2.args)))
        return applySubst(Subst(n => CFG(c1.kind, c1.name, [Var(ns[i]) for i in 1:l])), e)
    end
    return e
end

function commonSubst!(e::Exp, s1::Subst, s2::Subst)::Exp
    for (n1, e1) in s1, (n2, e2) in s1
        if (n1 != n2 && e1 == e2) && (s2[n1] == s2[n2])
            delete!(s1, n1)
            delete!(s2, n1)
            return applySubst(Subst(n1 => Var(n2)), e)
        end
    end
    return e
end

function msg(ng::NameGen, e1::Exp, e2::Exp)::Gen
    n = freshName(ng)
    s1 = Subst(n => e1)
    s2 = Subst(n => e2)
    e = Var(n)
    while true
        old_e = e
        e = commonSubst!(commonFunctor!(ng, e, s1, s2), s1, s2)
        (e == old_e) && break
    end
    return Gen(e, s1, s2)
end

export Gen, msg

end
